// @framer/ruucm.cursor-animation/code/Cursor.tsx
import * as React from "react";
import { Frame, useAnimation } from "framer";
import { useEffect } from "react";

const cursorWidth = 90;
const strokeWidth = 95;
export function Cursor() {
  const cursorControl = useAnimation();
  const handleClickEvent = event => cursorAnim(cursorControl, event);
  useEffect(() => {
    // for desktop click
    document.addEventListener("click", handleClickEvent, false);
    // for mobile touch (android)
    // document.addEventListener("touchstart", handleClickEvent);
    return () => {
      // for desktop click
      document.removeEventListener("click", handleClickEvent, false);
      // for mobile touch (android)
      // document.removeEventListener("touchstart", handleClickEvent);
    };
  }, []);
  return (
    <Frame
      initial={{
        scale: 0,
      }}
      animate={cursorControl}
      size={cursorWidth}
      background="transparent"
      radius={100}
      // use border as inline stroke
      border={strokeWidth / 2 + "px solid white"}
      style={{
        // use shadow as outline stroke
        boxShadow: `0px 0px 0px ${strokeWidth / 2}px rgba(255, 255, 255, 1),
        7px 7px 23px 0px rgba(0, 0, 0, 0.12)`,
      }}
    />
  );
}
const cursorAnim = async (control, event) => {
  // init values (redraw cursor)
  await control.start({
    scale: 0,
    boxShadow: `0px 0px 0px ${strokeWidth / 2}px rgba(255, 255, 255, 1),
      7px 7px 23px 0px rgba(0, 0, 0, 0.12)`,
    borderWidth: strokeWidth / 2,
    transition: {
      duration: 0,
    },
  });
  // move position
  control.start({
    top: event.pageY - cursorWidth / 2,
    left: event.pageX - cursorWidth / 2,
    transition: {
      duration: 0,
    },
  });
  // animate scale
  control.start({
    scale: 0.55,
    transition: {
      duration: 0.54,
    },
  });
  // animate strokes
  control.start({
    boxShadow: `0px 0px 0px 0px rgba(255, 255, 255, 1),
      0px 0px 0px 0px rgba(0, 0, 0, 0.12)`,
    borderWidth: 0,
    transition: {
      duration: 0.54,
      delay: 0.23,
    },
  });
};

Cursor.defaultProps = {
  width: 90,
  height: 90,
};
